### PR TITLE
Hostile mobs line of fire check

### PR DESCRIPTION
**Pull request by Sindorman**
This PR make few things:

- Hostile mobs now have a var `smart` which designates if they are smart enough to check their line of fire. In future I will also use this var to try and find path to their target around obstacles.

- If Hostile mob is smart and is a ranged mob, it will check it's line of fire before it shoots. If there is a friendly mob in line of fire, we will not shoot. If we hit some other object, but we still can hit our target penetrating that object, we will fire. If we can't hit our target, we just don't fire.

- Currently makes all hostile, ranged mobs smart.

- Cleans up hivebot code and improves it. Hivebot spawn beacon has now chance of spawning different kinds of bots each cycle, and has a spawn delay between each.

### CODE REVIEW

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 1]
```diff
@@ -81,10 +81,10 @@
 			else
 				to_chat(H, span("danger", "You sense great disturbance to your physical body!"))
 		else
-			user.visible_message("<span class='danger'>[user] attacked [src] with [I] waking [t_him] up!</span>", \
+			visible_message("<span class='danger'>[user] attacked [src] with [I] waking [t_him] up!</span>", \
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon_defense.dm]: Why did you remove the user from here? The second line is written with the user as the receiver, so this makes no sense now.

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 2]
```diff
@@ -81,10 +81,10 @@
...
 								"<span class='danger'>You attacked [src] with [I], but they do not respond... Maybe they have S.S.D?</span>")
 	else if(client && willfully_sleeping)
-		user.visible_message("<span class='danger'>[user] attacked [src] with [I] waking [t_him] up!</span>", \
+		visible_message("<span class='danger'>[user] attacked [src] with [I] waking [t_him] up!</span>", \
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon_defense.dm]: Same here as above.

File: code/modules/mob/living/simple_animal/hostile/hivebot.dm [hunk 3]
```diff
@@ -65,43 +63,48 @@
 	status_flags = 0
 	anchored = 1
 	stop_automated_movement = 1
-	var/bot_type = "norm"
+	var/bot_type = 1 // type of bot, 1 is normal, 2 is ranged, 3 is rapid rnaged.
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hivebot.dm]: Use a `#define`

File: code/modules/mob/living/simple_animal/hostile/hivebot.dm [hunk 4]
```diff
@@ -65,43 +63,48 @@
...
-		return
-
-
-	Life()
-		..()
-		if(stat == 0)
-			if(prob(2))//Might be a bit low, will mess with it likely
-				warpbots()
+
+/mob/living/simple_animal/hostile/hivebot/tele/New()
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hivebot.dm]: Initialize > New

File: code/modules/mob/living/simple_animal/hostile/hivebot.dm [hunk 5]
```diff
@@ -65,43 +63,48 @@
...
+	..()
+	var/datum/effect/effect/system/smoke_spread/smoke = new /datum/effect/effect/system/smoke_spread()
+	smoke.set_up(5, 0, src.loc)
+	smoke.start()
+	visible_message("<span class='danger'>The [src] warps in!</span>")
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hivebot.dm]:
```suggestion
	visible_message("<span class='danger'>\The [src] warps in!</span>")
```

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 6]
```diff
@@ -229,6 +234,23 @@ mob/living/simple_animal/hostile/hitby(atom/movable/AM as mob|obj,var/speed = TH
...
 	return
 
+/mob/living/simple_animal/hostile/proc/Check_fire(target_mob)
+	if(!target_mob)
+		return FALSE
+
+	var/target_hit = FALSE
+	var/friendly_hit = FALSE
+
+	for(var/atom/M in check_trajectory(target_mob, src, pass_flags=PASSTABLE))
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: You're doing two `istype` checks here. Just declare the variable of the loop as `var/A`.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 7]
```diff
@@ -229,6 +234,23 @@ mob/living/simple_animal/hostile/hitby(atom/movable/AM as mob|obj,var/speed = TH
...
+		if(ismob(M))
+			var/mob/L = M
+			if(L == target_mob)
+				target_hit = TRUE
+			if((L.faction == faction) || (L in friends))
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: Since this is n*m it might be wise to implement an early break after this.
```dm
if (friendly_hit && target_hit)
	break
```

Comment (Sindorman) [reply code/modules/mob/living/simple_animal/hostile/hostile.dm]: That won't do us much. putting break in that if statement is what we need. Idea is as soon as we hit friendly the whole thing is false.

Comment (Sindorman) [reply code/modules/mob/living/simple_animal/hostile/hostile.dm]: Actually we don't need hit_friendly bolean, I would just return FALSE as soon as we hit friendly.

File: code/modules/mob/living/simple_animal/hostile/hivebot.dm (hunk 4 above)
Comment (Sindorman) [reply code/modules/mob/living/simple_animal/hostile/hivebot.dm]: I only translated code into absolute pathing. I believe they use New as this behavior is intended to occur during beacon spawn.

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 8]
```diff
@@ -81,11 +81,13 @@
...
 		else
-			user.visible_message("<span class='danger'>[user] attacked [src] with [I] waking [t_him] up!</span>", \
-								"<span class='danger'>You attacked [src] with [I], but they do not respond... Maybe they have S.S.D?</span>")
+			visible_message("<span class='danger'>[user] attacked [src] with [I] waking [t_him] up!</span>")
+			if(user.client)
+				to_chat(user, "<span class='danger'>You attacked [src] with [I], but they do not respond... Maybe they have S.S.D?</span>")
 	else if(client && willfully_sleeping)
-		user.visible_message("<span class='danger'>[user] attacked [src] with [I] waking [t_him] up!</span>", \
-							"<span class='danger'>You attacked [src] with [I] waking [t_him] up!</span>")
+		visible_message("<span class='danger'>[user] attacked [src] with [I] waking [t_him] up!</span>",)
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon_defense.dm]: Why is this change necessary? The only thing potentially wrong about the old code is that `user` might be null, but if that is not the case, then the old code should work fine and is 100x more readable.

File: code/modules/mob/living/simple_animal/hostile/hivebot.dm [hunk 9]
```diff
@@ -65,43 +67,51 @@
...
+	var/datum/effect/effect/system/smoke_spread/smoke = new /datum/effect/effect/system/smoke_spread()
+	smoke.set_up(5, 0, src.loc)
+	smoke.start()
+	visible_message("<span class='danger'>\The [src] warps in!</span>")
+	playsound(src.loc, 'sound/effects/EMPulse.ogg', 25, 1)
+
+/mob/living/simple_animal/hostile/hivebot/tele/proc/warpbots()
+
+	if(!bot_amt)
+		qdel(src)
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hivebot.dm]: Psure you want to return after this.
```suggestion
		qdel(src)
		return
```

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 10]
```diff
@@ -229,6 +234,22 @@ mob/living/simple_animal/hostile/hitby(atom/movable/AM as mob|obj,var/speed = TH
 	target_mob = null
 	return
 
+/mob/living/simple_animal/hostile/proc/Check_fire(target_mob)
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: A bit of a minor quibble, but this is probably the weirdest naming style you can have. It's a mix of Pascal case and snake case. Please pick one or the other, not both.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 11]
```diff
@@ -229,6 +234,22 @@ mob/living/simple_animal/hostile/hitby(atom/movable/AM as mob|obj,var/speed = TH
...
 	return
 
+/mob/living/simple_animal/hostile/proc/Check_fire(target_mob)
+	if(!target_mob)
+		return FALSE
+
+	var/target_hit = FALSE
+
+	for(var/M in check_trajectory(target_mob, src, pass_flags=PASSTABLE))
+		if(ismob(M))
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: This is an unnecessarily complicated. `for (var/mob/M` would unwind to this exact same code. So please just type the for loop.

File: code/modules/mob/living/carbon/carbon_defense.dm (hunk 8 above)
Comment (Sindorman) [reply code/modules/mob/living/carbon/carbon_defense.dm]: That is the case, `user` becomes null. So it crashes and does get to lines which wake them up.

Comment (skull132) [reply code/modules/mob/living/carbon/carbon_defense.dm]: Then this fix will still crash. Because if `user == null` then doing `if (user.client)` will run to a null exception.

Keep the original line, since it is more concise, just add a check for user into the `else if`. So line 87 would become `else if (client && user && willfully_sleeping)`

Comment (Sindorman) [reply code/modules/mob/living/carbon/carbon_defense.dm]: Sorry. I meant when `user.client` is null. Hence attacking mobs with no client which are any simple mobs

Comment (skull132) [reply code/modules/mob/living/carbon/carbon_defense.dm]: This runtime doesn't make sense, then. `mob/visible_message` is safe to call even when the client of the mob is absent, thus, the original syntax is completely fine.

Can you post the actual stack of the runtime?
